File: src/ark_recognizer_py/recognize_icon.py
from dataclasses import dataclass

import cv2
from numpy.typing import NDArray

from ark_recognizer_py.utils import img_size


@dataclass
class IRResult:
    """
    Represents an icon recognition result.
    """
    x: float
    y: float
    radius: float


def estimate_circle_radius(scene_size: NDArray):
    w, h = scene_size
    # if the aspect ratio > 16:9, increasing the width will not increase the icon radius
    if w / h > 16 / 9:
        w = h * (16 / 9)
    # the icon radius is always 1/10 of the width
    estimate = w / 10
    # radius wanted, not diameter
    return estimate / 2


def recognize_icon(scene_image: NDArray):
    scene_image = cv2.cvtColor(scene_image, cv2.COLOR_BGR2GRAY)

    estimate = estimate_circle_radius(img_size(scene_image))
    estimate_low = round(estimate * 0.9)
    estimate_high = round(estimate * 1.1)
    circles = cv2.HoughCircles(scene_image,
                               cv2.HOUGH_GRADIENT, 1, round(estimate), param1=50, param2=30,
                               minRadius=estimate_low,
                               maxRadius=estimate_high)
    # cv2.HoughCircles() returns a 3D array.
    # however, the 1st dimension contains only 1 element and is completely redundant.
    circles = circles.reshape(-1, 3)
    # replace the radius from cv2.HoughCircles to estimated
    for circle in circles:
        circle[2] = estimate
    return circles

File: src/ark_recognizer_cpp/recognize_icon.cpp
//
// Created by Flandia on 2021/12/11.
//

#include "recognize_icon.h"
#include <vector>
#include <cmath>
#include <opencv2/imgproc.hpp>

using namespace std;
using namespace cv;

double estimate_circle_radius(const Size &size) {
    double w = size.width, h = size.height;
    if (w / h > 16.0 / 9.0) {
        w = h * (16.0 / 9.0);
    }
    double estimate = w / 10.0;
    return estimate / 2;
}


vector<IRResult> recognize_icon(Mat scene_image) {
    cvtColor(scene_image, scene_image, COLOR_BGR2GRAY);

    auto estimate = estimate_circle_radius(scene_image.size());
    auto estimate_low = (int) round(estimate * 0.9);
    auto estimate_high = (int) round(estimate * 1.1);
    vector<Vec3f> circles;
    HoughCircles(scene_image, circles,
                 HOUGH_GRADIENT, 1, estimate, 50, 30,
                 estimate_low, estimate_high);
    vector<IRResult> ir_results;

    std::transform(circles.begin(), circles.end(),
                   std::back_inserter(ir_results),
                   [&](Vec3f &circle) -> IRResult { return {circle[0], circle[1], estimate}; });
    return ir_results;
}
